### DISCUSSION TITLE
close long position at specific price

### DISCUSSION THREAD

**Discussion by berlinguyinca** [kernc/backtesting.py#228] [Q&A]
good morning. I would like to close a long position at a specific price. Say sell BTC at 40k. Right now the default price is calculated based on market price:

![image](https://user-images.githubusercontent.com/16364/105070685-7dca2400-5a38-11eb-824f-e1918a59dfff.png)

Which can lead to vastly and highly inflated wrong results, due to using 'Open' of the next candle.

I'm working on a trading bot, which uses internally different algorithms for automatic profit-taking on certain conditions. For example, close the positions when you hit 1%. 

so is there some way to provide 'position.close(price)' instead of just 'position.close()"

I have no interrest in the buy/sell, long/short approach.

**Comment by kernc** [comment]
You can use [`Backtest(..., trade_on_close=True)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest) which would execute `next()` right _before_ the closing of a bar, making your `position.close()` call execute with the closing price.

Alternatively, you can set a take-profit price for the [order](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Order.tp):
```py
self.buy(..., tp=40_000)
```
or for the existing trade by assigning to its [`.tp`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Trade.tp) property:
```py
trade = self.trades[0]
trade.tp = 40_000
```

**Reply by tombohub** [reply]
@kernc  I am trying to buy on Open and sell on Close the same candlestick, but it doesn't seem thats possible, is it?

```python
class Overnight(backtesting.Strategy):
    def init(self):
        self.buy()
       
    def next(self):
        self.position.close()
        self.buy()

bt = backtesting.Backtest(df, Overnight, trade_on_close=True)
stats = bt.run()
print(stats['_trades'])
```

It's now buying on Close and selling on Close next candlestick.

**Reply by replacementAI** [reply]
Can we please have a buy on open and sell on close feature? Its just not feasible with tp/sl.

**Comment by berlinguyinca** [comment]
yes sadly this does not seem to work for me :(

On Tue, Feb 9, 2021 at 4:55 PM tombohub <notifications@github.com> wrote:

> @kernc <https://github.com/kernc> I am trying to buy on Open and sell on
> Close the same candlestick, but it doesn't seem thats possible, is it?
>
> class Overnight(backtesting.Strategy):
>     def init(self):
>         self.buy()
>
>     def next(self):
>         self.position.close()
>         self.buy()
> bt = backtesting.Backtest(df, Overnight, trade_on_close=True)stats = bt.run()print(stats['_trades'])
>
> It's now buying on Close and selling on Close next candlestick.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kernc/backtesting.py/discussions/228#discussioncomment-354991>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAD73F4KXDQLFZAGHSCYRLS6HKPNANCNFSM4WJD2RAQ>
> .
>


-- 
------------------------------------------------------------
Lead Developer - Fiehnlab, UC Davis

gert wohlgemuth

work:
http://fiehnlab.ucdavis.edu/staff/wohlgemuth

linkedin:

https://www.linkedin.com/in/berlinguyinca

**Reply by tombohub** [reply]
@berlinguyinca  that's too bad.. Are you using some other backtester where you can open and close on the same candlestick?

**Comment by berlinguyinca** [comment]
No I'm using backtesting.py and have a simple wrapper around it to work
with my trading bot.

**Reply by tombohub** [reply]
@berlinguyinca can you buy at Open and close/sell at Close the same candlestick? 
How did you make it work?

**Comment by abhirathmahipal** [comment]
You can also try
```
self.sell(stop=x)
# or
self.buy(stop=x)
```
If the set price is between the high and low of the incoming candle, the order gets the price that you've specified.
